device/common/include/traccc/seeding/device/impl/count_grid_capacities.ipp

```diff
@@ -0,0 +1,58 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/seeding/spacepoint_binning_helper.hpp"
+
+// VecMem include(s).
+#include <vecmem/memory/device_atomic_ref.hpp>
+
+namespace traccc::device {
+
+TRACCC_HOST_DEVICE
+void count_grid_capacities(
+    std::size_t globalIndex, const seedfinder_config& config,
+    const sp_grid::axis_p0_type& phi_axis, const sp_grid::axis_p1_type& z_axis,
+    const spacepoint_container_const_view& spacepoints_view,
+    const vecmem::data::vector_view<const get_spacepoint_prefix_sum_element_t>&
+        sp_prefix_sum_view,
+    vecmem::data::vector_view<unsigned int> grid_capacities_view) {
+
+    // Check if anything needs to be done.
+    vecmem::device_vector<const get_spacepoint_prefix_sum_element_t>
+        sp_prefix_sum(sp_prefix_sum_view);
+    if (globalIndex >= sp_prefix_sum.size()) {
+        return;
+    }
+
+    // Get the spacepoint that we need to look at.
+    const get_spacepoint_prefix_sum_element_t sp_idx =
+        sp_prefix_sum[globalIndex];
+    device_spacepoint_const_container spacepoints(spacepoints_view);
+    const spacepoint sp = spacepoints.at(sp_idx);
```

Comment:
shouldn't const reference be better here?

Reply:
It is very good that you asked this question.

I don't have any measurements in these particular cases. But in the Acts code (https://github.com/acts-project/acts/tree/main/Plugins/Cuda/src/Seeding2) I did experiment with this just a little bit some time ago. I saw that pulling small objects into local (register) memory was usually a better idea compared to accessing them multiple times from global memory.

Since remember, when you use a (const) reference, it's only the pointer/reference that is in the thread's register. As soon as you try to access something from the referenced object, the thread has to go to global memory to access the info. Of course caching has improved a lot on GPUs in recent years, so in this case none of this might be important at all.

But I'm definitely standing on shaky ground with this... And unfortunately I just couldn't find the usernames of Nicolin on Xola, so let me pull them into this discussion in some other way... :thinking:

Reply:
for scalar variables const really does not make much of a difference. For CUDA we can have a look at the generated PTX code as NVCC will put it in a register anyway. PS: unless you need the extra range offered by 32 bit int you can do int16 which will save registers on Ampere and above GPUs as half precision ints will be packed into 32bit registers

Reply:
:thinking: Let's just go ahead like this for now. Probably it doesn't even make a difference whether we use references here or not, as the compiler may be smart enough to generate "the optimal" code in either case.

Doing this type of profiling of how well our code behaves is the exact sort of thing that I want Xola and Nicolin to help us with. So if later on we find that we should tweak this code, we will of course do it. :wink: